source/common/network/connection_impl.cc

```diff
@@ -468,6 +467,8 @@ void ConnectionImpl::onWriteReady() {
   } else if ((state_ & InternalState::CloseWithFlush) && new_buffer_size == 0) {
     ENVOY_CONN_LOG(debug, "write flush complete", *this);
     closeSocket(ConnectionEvent::LocalClose);
+  } else if (result.action_ == PostIoAction::KeepOpen && result.bytes_processed_ > 0) {
+    raiseEvent(ConnectionEvent::BytesSent);
```

Comment:
In certain cases, this may run a substantial amount of extra code callback code (high throughput streaming, TCP with no idle timeout, etc.). At minimum, I would suggest that this callback is opt-in. Basically, the connection would be told to raise it. This could be done in TCP proxy IFF there is an idle timeout. More fundamentally, for the TCP use case, I wonder if this is really even needed. Could idle timeout just be computed on receive from either upstream or downstream? Thoughts?

Reply:
The case I'm worried about is when there's a substantial bandwidth mismatch between the two sides.  If the fast side sent some large amount of data to envoy which is buffered, and envoy is slowly sending that to the other side, I don't want to idle-timeout because things are working fine and we're making progress.

For performance, that's a valid point.  However, for high-throughput it should scale reasonably well, because the write calls always try to write the entire buffer, so it should get pretty large chunks of data processed for each BytesSent event raised.  Although I think we do relatively small read chunks, so that could make it worse; but if that's the case, we should just read bigger chunks.

I'm hesitant to add an event to onEvent that is only used sometimes; I think that makes it harder to write filters.  I think a better approach might be to register a direct callback on the connection (which takes a std::function), and TcpProxy can do that when it needs to.  Thoughts on that?

Reply:
>The case I'm worried about is when there's a substantial bandwidth mismatch between the two sides. If the fast side sent some large amount of data to envoy which is buffered, and envoy is slowly sending that to the other side, I don't want to idle-timeout because things are working fine and we're making progress.

Yup that's fair. In flow control cases what you implemented is better.

> I think a better approach might be to register a direct callback on the connection (which takes a std::function), and TcpProxy can do that when it needs to. Thoughts on that?

Yup sure, I think opt-in for write callbacks sounds like a good idea to me. Why not even pass back the bytes written while you are at it? Then I think you can also kill the changes to HTTP connection managers.

Reply:
I agree with passing back the bytes written.

The thing in HTTP connection manager will be a real bug if anyone ever adds another event to onEvent, but I agree it won't make any sense as part of this change anymore.

Reply:
One other small note: Please think through what would happen if someone closes the connection in the callback. Someone might do this and it would be good if it didn't break. (This applies the same to the event version also).